src/config.py

```diff
@@ -215,29 +217,38 @@ def server_properties(self) -> List[str]:
         port = 9093 if self.charm.tls.enabled else 9092
         protocol = "SASL_SSL" if self.charm.tls.enabled else "SASL_PLAINTEXT"
 
+        rack_id = self.charm.model.get_relation(PEER).data[self.charm.unit].get(RACK_ID_KEY, None)
+
         properties = (
-            [
-                f"offsets.retention.minutes={self.charm.config['offsets-retention-minutes']}",
-                f"log.retention.hours={self.charm.config['log-retention-hours']}",
-                f"auto.create.topics={self.charm.config['auto-create-topics']}",
-                f"super.users={self.super_users}",
-                f"log.dirs={self.log_dirs}",
-                f"listeners={protocol}://:{port}",
-                f"advertised.listeners={protocol}://{host}:{port}",
-                f'listener.name.{(protocol).lower()}.scram-sha-512.sasl.jaas.config=org.apache.kafka.common.security.scram.ScramLoginModule required username="sync" password="{self.sync_password}";',
-                f"security.inter.broker.protocol={protocol}",
-                f"security.protocol={protocol}",
-            ]
-            + self.default_replication_properties
-            + self.auth_properties
-            + DEFAULT_CONFIG_OPTIONS.split("\n")
+                [
+                    f"offsets.retention.minutes={self.charm.config['offsets-retention-minutes']}",
+                    f"log.retention.hours={self.charm.config['log-retention-hours']}",
+                    f"auto.create.topics={self.charm.config['auto-create-topics']}",
+                    f"super.users={self.super_users}",
+                    f"log.dirs={self.log_dirs}",
+                    f"listeners={protocol}://:{port}",
+                    f"advertised.listeners={protocol}://{host}:{port}",
+                    f'listener.name.{(protocol).lower()}.scram-sha-512.sasl.jaas.config=org.apache.kafka.common.security.scram.ScramLoginModule required username="sync" password="{self.sync_password}";',
+                    f"security.inter.broker.protocol={protocol}",
+                    f"security.protocol={protocol}",
+                ]
+                + ([f"rack.id={rack_id}"] if rack_id is not None else [])
+                + self.default_replication_properties
+                + self.auth_properties
+                + DEFAULT_CONFIG_OPTIONS.split("\n")
         )
 
         if self.charm.tls.enabled:
             properties += self.tls_properties
 
         return properties
 
+    def set_rack_id(self, rack_id: str) -> 'KafkaConfig':
+        relation = self.charm.peer_relation
+        if relation is not None:
+            relation.data[self.charm.unit].update({RACK_ID_KEY: rack_id})
+        return self
```

Comment:
I actually started to implement things the way you suggest, that was somewhat along the line of `set_secret` and so on. But I wasn't too happy. The reason for that was that in order to keep the name of the databag field (rack-id) in a single place, I created a constant in config.py (the `RACK_ID_KEY`) and then import that in charm.py. But then I started to think, why not centralizing IO operation (for KafkaConfig related field) in a single module/class. I knew you would notice :D 

Note that also in this implementation the information is always stored in the peer relation data (so that does not change), and there is already link to the peer realtion data for KafkaConfig. Basically, now KafkaConfig only exposes read operation. Why not also exposing writing also to keep business logic in a single module instead of scattered (reading in confing and writing in charm)? I'd personally centralize IO operation (for KafkaConfig related field) into a single component, that acts as a proxy to peer relation data

Reply:
I would like to keep `KafkaConfig` (and any other class that does not inherit from `ops.framework.Object` with observed events, e.g `KafkaAuth`) not being responsible for the setting of relation data. From a charm-dev POV, it's a helpful model of 'observed event hooks set state to peer data' with 'workload tasks/config refer to current state', especially when handling out-of-order events fired from Juju. Relation 'state' data persists, instance state is lost on each event fire.

I can see how it makes sense to have all the 'config' related stuff within a single class. You're right in that it's also a leap of knowledge to know that `key=rack-id` set in one file, is then used to get `rack-id=value` in another, but I think that problem is perhaps better addressed with building some consistent schema object (mentioned in our other comment thread) down the line.